### PR TITLE
fix: size and validate the CVC from the detected card brand

### LINKED ISSUE #1706: CVC input accepts 4 digits regardless of card brand
## TL;DR

The CVC input hardcodes `maxLength=4` in both the combined card element and the standalone `cardCvc` element, so a 4-digit CVC is accepted on a Visa and the payment is only rejected by the processor, as an opaque `cvv_invalid`.

## Where

Against [eee6915](https://github.com/juspay/hyperswitch-web/commit/eee69151):

- [`CardCVCElement.res:245`](https://github.com/juspay/hyperswitch-web/blob/eee69151/src/CardCVCElement.res#L245) (individual element)
- [`CardPayment.res:666`](https://github.com/juspay/hyperswitch-web/blob/eee69151/src/Payments/CardPayment.res#L666) (combined element)

The per-brand lengths already exist. [`CardPattern.res`](https://github.com/juspay/hyperswitch-web/blob/eee69151/shared-code/sdk-utils/validation/CardPattern.res) declares `maxCVCLength` (4 for American Express and Maestro, 3 for most others, 4 for the default pattern), and `cvcNumberInRange` validates against it. Only the input attribute was left literal.

## Reproduce

Enter `4242 4242 4242 4242` with a 4-digit CVC and confirm. The field accepts the fourth digit; the processor returns `cvv_invalid`, which does not indicate the field or the reason.

Note that `maxLength` constrains only new keystrokes, so a CVC entered before the card number, or pasted, can exceed the brand's limit even once the brand is known.

## Note for the individual elements

They are one per iframe, so `cardCvc` never sees the PAN and cannot detect the brand on its own; constraining it there needs the brand from `cardNumber`.

### PR DESCRIPTION
**Pull request by hossam-magdy**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Fixes #1706.

In a bare `card` element every CVC rule runs against an empty brand, so the rules fall back to the default card pattern, which accepts 3 or 4 digits, and the input's `maxLength` was a literal `4`. A 4-digit CVC on a Visa therefore passed the client and failed at the processor as an opaque `cvv_invalid`.

This derives a CVC-scoped brand and routes the CVC rules through it: the input's `maxLength`, the formatter, the range checks, and the CVC-to-ZIP auto-advance. A CVC already in the field when the brand arrives is re-judged against it, by the same rule a blur already uses, so narrowing invalidates it and widening makes it valid again. The value the user typed is never rewritten and never deleted.

The hook's `cardBrand` is left byte-identical to upstream. That is deliberate and is the whole reason the change is scoped this way: widening it would also put `card_network` on the wire in this mode, make `checkIsCardSupported` decline brands outside the merchant's configured `card_networks`, narrow `maxCardLength`, and wake the clear-expiry-and-CVC-on-brand-change effect. A PR about CVC length should change none of those, and this one changes none of them. `Payment.res` is untouched.

### Three things reviewers should see explicitly

- **The individual `cardCvc` element is NOT fixed.** It is a separate iframe whose local card number is always empty, so it cannot know the brand, and there is no association between a `cardNumber` and the `cardCvc` of the same form to route one to it. Its input keeps `maxLength=4`. Tightening only the submit-time gate was tried and dropped: it rejects input that the CVC iframe still shows as valid, and the error surfaces in a frame with no CVC field, which is worse than the status quo. Closing this properly looks like a design decision rather than something to invent inside a bugfix.
- **The two saved-card CVC inputs keep `maxLength=4` on purpose.** In those flows the scheme atom IS written, so the CVC brand is the saved card's scheme and the formatter already caps the value. Measured: typing `1234` into a saved Visa's CVC re-collect field yields `123`, with the attribute still reading 4. The literal never binds, so converting it would be churn.
- **In the payment element the `maxLength` attribute does change** (4 becomes 3 on a Visa), because that is where the brand was already available. The submitted value does not change, since the formatter already capped it. The only user-visible delta is that the browser now blocks a phantom 4th keystroke it previously accepted and discarded.

### One pre-existing gap this does not widen, but worth knowing

`emitIsFormReadyForSubmission` only fires when all three of `(isCardValid, isExpiryValid, isCVCValid)` are `Some`, so any transition to `None` emits nothing and a previously-emitted `true` is never retracted. Upstream already reaches that state on its own: `changeCVCNumber` sets `None` on any invalid keystroke, so deleting a digit from an accepted CVC leaves the same stale `true`. Using the blur rule for the re-judge means this change reaches `Some(false)` rather than `None` when a CVC goes stale, so the signal retracts correctly here. Making the emit total is a change to the merchant-facing event contract and belongs in its own PR.

## How did you test it?

Added two cypress assertions to `02-cards/01-card-validation.cy.ts`, next to the existing per-brand formatting tests: the CVC `maxlength` is 3 for a Visa and 4 for an American Express. The first fails on `main`. Note the `cypress` job is skipped on fork PRs, so these have not run in CI here.

Beyond that, manually against a hosted sandbox with a real connector, driving the elements in a browser. The bare `card` element is not in any demo UI, so it was mounted directly to exercise it.

CVC `maxlength`:

| | before | after |
|---|---|---|
| bare `card` element, no PAN | 4 | 4 |
| bare `card` element, Visa | 4 | 3 |
| bare `card` element, Amex | 4 | 4 |
| payment element, Visa / Amex | 4 / 4 | 3 / 4 |
| individual `cardCvc` element | 4 | 4 (by design, see above) |
| saved-card CVC re-collect | 4 | 4 (value already capped, see above) |

Validity and submit, in the bare `card` element. These matter more than the attribute, because a wrong-length CVC used to be reported as valid:

| Case | Result |
|---|---|
| Amex PAN, CVC `123`, blur | invalid, and form readiness emits `false` |
| Visa PAN, CVC `123`, blur | valid, readiness `true` |
| Amex PAN, CVC `1234`, blur | valid, readiness `true` |
| CVC `1234` typed before the PAN, then a Visa PAN | value `1234` kept, marked invalid |

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [x] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by XyneSpaces** [comment]
[should-fix] The new `messageParentWindow` call in `src/RenderPaymentMethods.res` posts the detected card brand using the default `targetOrigin="*"`. Please pass the validated parent origin from the `keys` atom instead, so this brand message is only delivered to the expected merchant origin.

```suggestion
// src/RenderPaymentMethods.res
let {parentURL} = Recoil.useRecoilValueFromAtom(keys)
...
switch paymentType {
| CardNumberElement =>
  messageParentWindow(
    [("cardBrandForSiblings", detectedBrand->JSON.Encode.string)],
    ~targetOrigin=parentURL,
  )
| _ => ()
}
```

**Comment by XyneSpaces** [comment]
[should-fix] `src/RenderPaymentMethods.res`: new `cardBrandForSiblings` postMessage uses a wildcard origin

`messageParentWindow` defaults to `targetOrigin="*"`. Other postMessage callers in the codebase (e.g., `InputField.res`) pass `~targetOrigin=parentURL` from the `keys` atom. Broadcast the brand update with the validated parent origin instead of `*`.

Also confirm that `Elements.res` fan-out via `iframePostMessage` targets a constrained origin (it currently routes through `GlobalVars.targetOrigin`) and is not `"*"`.

**Comment by hossam-magdy** [comment]
Thanks for both notes (https://github.com/juspay/hyperswitch-web/pull/1707#issuecomment-5081450838 and https://github.com/juspay/hyperswitch-web/pull/1707#issuecomment-5081665667).

Both point at `src/RenderPaymentMethods.res` posting a `cardBrandForSiblings` message, which was in the revision that was on this PR when you commented. That approach (publishing the brand from the `cardNumber` frame and relaying it to its siblings) has since been dropped entirely and the branch force-pushed. The current head touches four files and adds no postMessage, no cross-frame message and no origin-bearing call at all, so there is nothing left here to constrain.

On the second half, about the `Elements.res` fan-out: the answer is the opposite of the confirmation you were after.

- `GlobalVars.targetOrigin` is a hardcoded wildcard: https://github.com/juspay/hyperswitch-web/blob/eee6915192960204c438effdf3049a0d5d31c4ac/src/GlobalVars.res#L13
- `messageParentWindow` defaults to the same, and most call sites take the default: https://github.com/juspay/hyperswitch-web/blob/eee6915192960204c438effdf3049a0d5d31c4ac/src/Utilities/Utils.res#L38
- `keys.parentURL` is also `"*"` by default, so passing it would not have narrowed anything either: https://github.com/juspay/hyperswitch-web/blob/eee6915192960204c438effdf3049a0d5d31c4ac/src/Hooks/CommonHooks.res#L191

That is existing behaviour rather than something this PR introduces, so I have deliberately kept it out of a CVC-length change. Happy to open a separate issue for it if that would be useful.

File: src/Hooks/CommonCardProps.res [hunk 1]
```diff
@@ -50,7 +50,16 @@ let useCardForm = (~logger, ~paymentType) => {
...
+  let detectedCardBrand = cardBrand
   let cardBrand = CardUtils.getCardBrandFromStates(cardBrand, cardScheme, showPaymentMethodsScreen)
+
+  // Nothing in the card element's tree writes the scheme atom, so the brand is lost here and the CVC
+  // rules fall back to the permissive default. CVC-scoped: widening cardBrand changes the payload.
+  let cardBrandForCvc = switch (cardBrand, paymentType) {
+  | ("", CardThemeType.Card) => detectedCardBrand
+  | _ => cardBrand
+  }
+  let maxCVCLength = CardValidations.getobjFromCardPattern(cardBrandForCvc).maxCVCLength
```
Comment (aritro2002) [inline src/Hooks/CommonCardProps.res:74]: why do we need `detectedCardBrand`? why can't we use `cardBrand` directly?

**Comment by aritro2002** [comment]
https://github.com/user-attachments/assets/91786720-b398-48ea-a8dc-750d5f9a7bd7

I noticed that when I focus on the CVC input and then blur it, the colour does not change to red.

**Comment by aritro2002** [comment]
Could you also please attach videos showing the behaviour before and after your changes?

**Review by hossam-magdy — COMMENTED** [review]
**Re:** [this comment](https://github.com/juspay/hyperswitch-web/pull/1707#issuecomment-5090316404)

Thanks for the catch. Reproduced and fixed in [0aa3dae](https://github.com/juspay/hyperswitch-web/commit/0aa3daed2abf144bc6bb1cd7eb2453bc29429330).

The cause is the auto-advance rather than the blur. Completing the expiry focuses the CVC ([CommonCardProps.res#L211](https://github.com/juspay/hyperswitch-web/blob/146f65fa289bb809a3ce6268c68bcc8d51a642c0/src/Hooks/CommonCardProps.res#L211)), and focusing clears that field's validity ([InputField.res#L72](https://github.com/juspay/hyperswitch-web/blob/146f65fa289bb809a3ce6268c68bcc8d51a642c0/src/Components/InputField.res#L72)), so the verdict went without the user editing anything. It now auto-advances only when the CVC is empty.

Submit stayed blocked throughout ([Payment.res#L56](https://github.com/juspay/hyperswitch-web/blob/146f65fa289bb809a3ce6268c68bcc8d51a642c0/src/Payment.res#L56)), so the defect was the missing red only.

File: src/Hooks/CommonCardProps.res (hunk 1 above)
Comment (hossam-magdy) [reply src/Hooks/CommonCardProps.res:74]: Because `cardBrand` is `""` here. One line above it is reassigned from [`getCardBrandFromStates`](https://github.com/juspay/hyperswitch-web/blob/146f65fa289bb809a3ce6268c68bcc8d51a642c0/src/CardUtils.res#L765), which returns the `cardBrand` atom when `showPaymentMethodsScreen` is false, and that atom is only ever written in [PaymentManagement.res#L17](https://github.com/juspay/hyperswitch-web/blob/146f65fa289bb809a3ce6268c68bcc8d51a642c0/src/PaymentManagement.res#L17).

The root fix would be `getCardBrandFromStates` itself. Kept CVC-scoped because widening `cardBrand` also moves `maxCardLength`, `checkIsCardSupported`, `cardType` and the `card_network` in the confirm payload, among others. Happy to go that way instead, if that is deemed best.

File: src/Hooks/CommonCardProps.res [hunk 2]
```diff
@@ -208,7 +231,11 @@ let useCardForm = (~logger, ~paymentType) => {
     logInputChangeInfo("cardExpiry", logger)
     let formattedExpiry = val->CardValidations.formatCardExpiryNumber
     if isExipryValid(formattedExpiry) {
-      handleInputFocus(~currentRef=expiryRef, ~destinationRef=cvcRef)
+      // Focusing clears a field's validity, so auto-advancing into an already-filled CVC would hide
+      // the verdict the brand re-judge just set. Only pull focus when there is nothing to hide.
+      if cvcNumber->String.length == 0 {
```
Comment (aritro2002) [inline src/Hooks/CommonCardProps.res]: Normally, when a user finishes typing the expiry date, the cursor should automatically jump to the CVC field. This helps users move through the form quickly without clicking each field manually.
With this change, that auto-jump is blocked whenever the CVC field already has a value.

File: src/Hooks/CommonCardProps.res (hunk 1 above)
Comment (aritro2002) [reply src/Hooks/CommonCardProps.res:74]: okay got it

File: src/Hooks/CommonCardProps.res (hunk 2 above)
Comment (aritro2002) [reply src/Hooks/CommonCardProps.res]: I’m still seeing the same behaviour. When I focus on the CVC input and then blur it while the value is invalid, the colour does not change to red.

Comment (hossam-magdy) [reply src/Hooks/CommonCardProps.res]: Thanks @aritro2002 . That's the kind of edge cases that are well known to maintainers off top of their heads. Appreciate the input.

I've tested now all the cases raised and came on mind. Below is a recording:


https://github.com/user-attachments/assets/f342830f-223b-42f6-997a-eae1b1825f68


---


[As a side note]: perhaps a round of automated tests could have caught these issues already and saved some manual testing time (on both sides). Though, that's probably a conversation of a different scope.

---

Please do let me know if there's an/another edge cases that we/I could have missed.

Thanks again

Comment (hossam-magdy) [reply src/Hooks/CommonCardProps.res]: > [As a side note]: perhaps a round of automated tests could have caught these issues already and saved some manual testing time (on both sides). Though, that's probably a conversation of a different scope.

Correcting myself. There's already a set in `cypress-tests/cypress/e2e/`. I'll check leveraging it …

Comment (hossam-magdy) [reply src/Hooks/CommonCardProps.res]: Added a couple of cypress tests ✅

**Comment by hossam-magdy** [comment]
Any suggestions how to move this PR forward / make a decision on it? … cc: @aritro2002 @Shivam25092001

**Comment by Shivam25092001** [comment]
@AbhishekChorotiya @sakksham7 could you please help us with the review here.

**Review by aritro2002 — DISMISSED** [review]

**Review by ArushKapoorJuspay — APPROVED** [review]
